qlib/Logger.qm

```diff
@@ -0,0 +1,2717 @@
+        */
+        public process(timeout ms = -1) {
+            # normally function is called from one thread and it is not time critical
+            AutoLock al(m_lock);
+            # get new events and match with pending
+            hash lastMatch;
+            while (True) {
+                *hash e = getEvent(ms);
+                if (!e) {
+                    break;
+                }
+                if (lastMatch.appender != e.appender) {
+                    bool found = False;
+                    foreach string id in (keys m_pending) {
+                        if (m_pending{id}.appender == e.appender) {
+                            lastMatch.id = id;
+                            lastMatch.appender = e.appender;
+                            found = True;
+                            break;
+                        }
+                    }
+                    if (!found) {
+                        # look into processing
+                        foreach string id in (keys m_processing) {
+                            if (m_processing{id}.appender == e.appender) {
+                                lastMatch.id = id;
+                                found = True;
+                                break;
+                            }
+                        }
+                        if (!found) {
+                            # add new entry
+                            lastMatch.id = string(m_lastId.next());
+                        }
+                        lastMatch.appender = e.appender;
+                        m_pending{lastMatch.id} = (
+                            "id": lastMatch.id,  # BUG: see ThreadPool::submit(id, ...) call
+                            "appender": e.appender,
+                            "events": (),
+                        );
+                    }
+                }
+                push m_pending{lastMatch.id}.events, e.('type', 'params');
+            }
+
+            # remove finished
+            while (m_finished.size() > 0) {
+                string id = m_finished.get(-1);
+                remove m_processing{id};
+            }
+            # now try push pending
+            foreach string id in (keys m_pending) {
+                # remove finished
+                #while (m_finished.size() > 0) {
+                #    remove m_processing{m_finished.get(-1)};
+                #}
+                # is available a free worker thread ?
+                if (m_maxThreads >= 0 && m_processing.size() >= m_maxThreads) {
+                    return;
+                }
+                if (exists m_processing{id}) {
+                    # we cannot trigger new thread if is being processed
+                    continue;
+                }
+                hash e = remove m_pending{id};
+                m_processing{id} = e;
+                # BUG: when passing "id" then no copy is passed to submit pool queue and worker may get wrong value
```

Comment:
I don't understand the bug here, please explain - also what does "real constant enough value" mean?